### DISCUSSION TITLE
Grab URL if front-most-facing application is a Browser on macOS

### DISCUSSION THREAD

**Discussion by dougpagani** [espanso/espanso#1249] [Feature requests and Ideas]
**Is your feature request related to a problem? Please describe.**
I'm not sure this is a problem or if it's abnormal behavior, but on macos, from `espanso detect`, I get:
![image](https://user-images.githubusercontent.com/18250753/148669405-9b495688-43da-4640-a423-3c239f651cb6.png)

If you'll notice, the bundle id is used for the window title. This is the same for all apps I've tried w/ `espanso detect` running.

-------
**Describe the solution you'd like**
There are two additional pieces of information that could be garnered from macos:
1. Title, for all apps
2. URL, if it is a browser

### 1. Title, for all apps
I'm not sure if there's a reason why this hasn't been developed, but this info is accessible via `kCGWindowName` in macos:
![image](https://user-images.githubusercontent.com/18250753/148669436-cb07c5ed-2076-487f-959f-1bed34654743.png)

_... which will yield something like:_
`windowTitle: 'Unicorns - Google Search'`

### 2. URL, if it is a browser
And then for URL, that can be obtained via `osascript` (CAVEAT: this is not a fast call like `kCGWindowName` afaik):
![image](https://user-images.githubusercontent.com/18250753/148669501-080ec94f-cdac-4ef8-9374-4203e1a0559f.png)
_... which will yield something like:_
`url: 'https://sindresorhus.com/unicorn'`

------
**Describe alternatives you've considered**
n/a

-------
**Additional context**
This would add value since we're moving towards a world of webapps where the front-most app is virtually indistinguishable from the point of view of bundle-id's and such, but is really quite different.

**Comment by federico-terzi** [comment]
Hey @dougpagani,

Thank you for the feedback!

> 1. Title, for all apps

The new v2 introduced support for the window titles that was missing in the previous version, so we could consider this one solved :)

> 2. URL, if it is a browser
> And then for URL, that can be obtained via osascript (CAVEAT: this is not a fast call like kCGWindowName afaik):

Thanks! This is interesting. Given that Espanso polls the active application for this information (exec, title, class) approximately every second, this approach might have a negative impact on performance, so we should think about a way to mitigate the overhead. Perhaps, we could scan the URL only if certain conditions are met (like the window title being equal to something), but that would possibly require a significant change to the detection logic :)

> This would add value since we're moving towards a world of webapps where the front-most app is virtually indistinguishable from the point of view of bundle-id's and such, but is really quite different.

Do you think having the window title might solve this problem?

Cheers :)

**Comment by dougpagani** [comment]
Hi @federico-terzi  _(This post is purely for your knowledge, and so to save you some time, I time-linked the video -- it's a quick 1-minute show of the steno system's macros. I know you're in "stepping away" mode with the project, but I figured I'd close a good conversation now that I found a very high-efficiency analogue demo for you as a matter of general knowledge.)_

, I can't find the other conversation we had a while ago where I had asked for a mechanism to be able to back-query the last expansion - and you asked something like _"can you show/explain an example of why that would be useful?"_

I just came across a very succinct demo of what I think is mostly perfect design of handling of "User-Bankable apps", which Espanso is very much so one of, in this video on a steno system:

https://youtu.be/hLgf-Zkte48?si=INQ4WdNijuYog8ri&t=1m

BANKINGS META COMMANDS:
        - 1: FOCUS the app
        - 2: ADD translation
            - If untranslated-stroke was entered, then it prompts for a translation. (Prefilling the stroke)
            - If finger-spelled, then it prompts for a stroke. (Prefilling the translation)
        - 3: SEARCH translations 
        - 4: TOGGLE active 

In my opinion, to satisfy "fluent self-growth", it is only missing an "EDIT translation", which could just be overloaded behavior of "ADD", such that the last stroke is opened & hoisted for fixing.

I've built a lot of applets like Espanso, and I've found that the highest bandwidth comes from when you have a context-aware `(ADD/EDIT)-FROM-LAST` command, when possible with a "mappings-bindings candidate log" (which Espanso has by virtue of needing to resolve triggers to expansions).

I wanted to expose the "last candidate" for an expansion because this would enable an ADD-FROM-LAST. I do this for virtually everything and it makes for good software design, especially software meant for other software developers. "alias-last()" on the shell would be another easily understood example of this.

Sure, the key logging thing is somewhat of a concern, but it only expands the scope of memory from 1 word to 2, so I really don't think it's that much more of a security risk. If you could query the last trigger from the CLI-tool, you could easily write a system shortcut to pull up & navigate to the YAML entry & quickly edit it. Would reduce expansion-iteration time from ~1 minute per fixup to ~2 seconds. And extension/growth of the expansions bank would be likewise, incredibly easy, but this doesn't require much help from Espanso if it isn't built-in, as you can just use clipboard content + zenity-GUI-prompt-script which prefills the expansion and prompts for you to fill the trigger (then auto-formatting the YAML entry and dumping it into a valid search-path file as appended text).
